[Merged by Bors] - feat(measure_theory/group/add_circle): fundamental domain for rational angle rotations in the circle

ocfnash opened the pull request:
The main results are `add_circle.is_add_fundamental_domain_of_ae_ball` and its corollary `add_circle.volume_of_add_preimage_eq`.

---

- [x] depends on: #17321
- [x] depends on: #17433
- [x] depends on: #17438
- [x] depends on: #17590

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

ocfnash committed:
feat(analysis/normed/group/add_circle): results about finite-order points on the circle

ocfnash committed:
Merge branch 'master' into ocfnash/circle_finite_order

Hoping to fix a deterministic timeout unrelated to my work

ocfnash committed:
Fix lint

ocfnash committed:
feat(group_theory/order_of_element): lemmas about orders of powers of group elements

The lemma `subgroup.mk_eq_one_iff` is unrelated but convenient to include
for dependent work.

ocfnash committed:
feat(measure_theory/group/fundamental_domain): add lemma `measure_theory.is_fundamental_domain.mk_of_measure_univ_le`

Also some supporting / related lemmas.

ocfnash committed:
Make argument explicit in `order_eq_card_zpowers'`

ocfnash committed:
Merge branch 'ocfnash/circle_finite_order' into ocfnash/circle_ball_fd

ocfnash committed:
Merge branch 'ocfnash/fin_order_lemmas' into ocfnash/circle_ball_fd

ocfnash committed:
feat(is_add_fundamental_domain_of_ae_ball): fundamental domain for rational angle rotations in the circle

ocfnash committed:
Fix build

ocfnash committed:
Fix lint (missing doc string)

ocfnash committed:
merge master

ocfnash committed:
Fixes after merging master

ocfnash committed:
Finish fixing after merging master

ocfnash committed:
Drop `set.has_smul_pi_Prop`

Better just to use type ascriptions

sgouezel commented on src/measure_theory/group/add_circle.lean:
```diff
@@ -0,0 +1,107 @@
+variables {T : ℝ} [hT : fact (0 < T)]
+include hT
+
+lemma closed_ball_ae_eq_ball {x : add_circle T} {ε : ℝ} :
```
You shouldn't need `hT` for this one.

ocfnash replied:
Agreed. However unfortunately `add_circle.measure_space` requires `hT` so I'd have to change that instance to implement this.

I'm happy to do this if you think I should?

sgouezel replied:
Indeed, `hT` should not be needed to define the measure on the quotient (we're quotienting the real line by a discrete subgroup, so the measure on the quotient always makes sense). Ideally, we shouldn't even need to discuss whether T is zero or not in the definition of the measure...

ocfnash replied:
Hmm, I don't see an easy way to avoid branching on whether `T` is zero (though I agree this is desirable).

I suppose one approach is to change [this definition](https://github.com/leanprover-community/mathlib/blob/3891ab62ecddc54174f19188fd850a5d6f3cfdde/src/measure_theory/integral/periodic.lean#L59) to:
```lean
noncomputable instance measure_space : measure_space (add_circle T) :=
{ volume := (if T = 0 then volume else volume.restrict $ Ioc 0 (|T|)).map coe,
  .. add_circle.measurable_space }
```
and then prove:
```lean
variables [hT : fact (0 < T)]
include hT

/-- When `0 < T`, the volume on `add_circle T` is just `T` times the Haar measure. -/
lemma volume_eq : (volume : measure $ add_circle T) = ennreal.of_real T • add_haar_measure ⊤ :=
begin
  change (if T = 0 then volume else volume.restrict $ Ioc 0 (|T|)).map coe = _,
  simp only [hT.out.ne.symm, if_false, abs_eq_self.mpr hT.out.le],
  letI : measurable_space (ℝ ⧸ zmultiples T) := add_circle.measurable_space,
  letI : borel_space (ℝ ⧸ zmultiples T) := add_circle.borel_space,
  letI : compact_space (ℝ ⧸ zmultiples T) := add_circle.compact_space T,
  have := (is_add_fundamental_domain_Ioc' hT.out 0).map_restrict_quotient ⊤ (by simp),
  simp only [zero_add] at this,
  erw this,
  simp,
end
```
and proceed from there.

What do you think?

sgouezel replied:
Sounds great. Or, in the definition, you can even use `ennreal.of_real T • add_haar_measure ⊤` when `T` is positive, after registering the necessary instances, depending on what is more convenient for you.
Then you can register the instance that `volume` is an `add_haar_measure` in general.

ocfnash replied:
This is turning out to be non-trivial because quite a few facts for the circle are proved for `0 < T` rather than `T ≠ 0`. In several cases, we would need their generalisation not just for proofs but even for statements. For example We only know the circle is compact when `0 < T` and so we cannot talk about `add_haar_measure ⊤` when `T <  0`.

After experimenting with a few approaches I think the only clean way to do things would be to replace the `fact (0 < T)` pattern entirely with `fact (T ≠ 0)`. (I admit a bias here -- I did not actually introduce the `fact (0 < T)` pattern and have always harboured some reservations). I am willing to do this but I would like to make the case that it should happen in a follow-up PR (which would then also address the point here about `closed_ball_ae_eq_ball`).

I have considered alternative approaches including ingredients such as:

- allowing a mixture of `fact (0 < T)` and `fact (T ≠ 0)` hypotheses
- attempt to stick with `fact (0 < T)` throughout and arguing using a measure-preserving map `add_circle T → add_circle (-T)`
- directly attack the problem that we have two names for the same subgroup (`zmultiples T`, `zmultiples (-T)`) and thus two different types `add_circle T` and `add_circle (-T)`. E.g., redefining `add_circle` to consume a subgroup together with a discreteness hypotheses or a proof that it is set-wise `zmultiples T` for some `T`

I have mostly convinced myself that these really just push the mess around and the best solution is to migrate to `fact (T ≠ 0)` but I have fully convinced myself that it is reasonable for this work to take place in a separate PR. (Nevertheless, I welcome dissenting opinions.)

What do you think?

sgouezel replied:
I agree that this is better for a followup PR. You can merge this one once #17590 is in.
bors d+

sgouezel commented on src/measure_theory/group/add_circle.lean:
```diff
@@ -0,0 +1,107 @@
+    exact measure_mono (closed_ball_subset_ball hr.2), },
+end
+
+lemma is_add_fundamental_domain_of_ae_ball (I : set $ add_circle T)
```
Can you add a docstring to this lemma?

ocfnash replied:
Done.

sgouezel commented on src/measure_theory/group/add_circle.lean:
```diff
@@ -0,0 +1,107 @@
+  (u x : add_circle T) (hu : is_of_fin_add_order u)
+  (hI : I =ᵐ[volume] ball x (T / (2 * add_order_of u))) :
+  is_add_fundamental_domain (add_subgroup.zmultiples u) I :=
+is_add_fundamental_domain.mk_of_measure_univ_le
```
It's a little bit weird to mix term mode and tactic mode like that, with two big `begin ... end` blocks. Would you go full tactic mode?

ocfnash replied:
Done, and it did indeed improve readability.

sgouezel commented on src/measure_theory/group/measure.lean:
```diff
@@ -103,6 +104,18 @@ lemma measure_preserving.mul_right (μ : measure G) [is_mul_right_invariant μ]
   measure_preserving (λ x, f x * g) μ' μ :=
 (measure_preserving_mul_right μ g).comp hf
 
+@[to_additive]
```
This will probably clash with #17590...

ocfnash replied:
Oh indeed, what a striking coincidence. I've added #17590 as a dependency to this PR. I'll merge master in here once it lands and deduplicate etc.

ocfnash committed:
Suggestion from CR: use full tactic mode in proof of `is_add_fundamental_domain_of_ae_ball` and add doc string.

ocfnash committed:
Merge branch 'master' into ocfnash/circle_ball_fd

ocfnash committed:
Fix after merging master.

Having done this, I now realise that this PR should have been a dependency
for #17590 rather than the other way around.

ocfnash committed:
Merge fix: drop duplicated import
